packages/aws-cdk/lib/api/util/checks.ts

```diff
@@ -0,0 +1,88 @@
+import { debug } from '../../logging';
+import { ISDK } from '../aws-auth/sdk';
+
+export async function determineAllowCrossAccountAssetPublishing(sdk: ISDK, customStackName?: string): Promise<boolean> {
+  const stackName = customStackName || 'CDKToolkit';
+  if (!(await _stagingBucketInToolkitStack(sdk, stackName))) {
+    // indicates an intentional cross account setup
+    return true;
+  }
+
+  if (await _bootstrapVersion(sdk, stackName) >= 21) {
+    // bootstrap stack version 21 contains a fix that will prevent cross
+    // account publishing on the IAM level
+    // https://github.com/aws/aws-cdk/pull/30823
+    return true;
+  }
+
+  // other scenarios are highly irregular and potentially dangerous so we prevent it by
+  // instructing cdk-assets to detect foriegn bucket ownership and reject.
+  return false;
+}
+
+async function _stagingBucketInToolkitStack(sdk: ISDK, stackName?: string): Promise<boolean> {
+  const toolkitStackName = stackName || 'CDKToolkit';
+  try {
+    // Create the SDKs for this enviornment, and use it to get the stack
+    const cfn = sdk.cloudFormation();
+    const s3 = sdk.s3();
+    const response = await cfn.describeStacks({ StackName: toolkitStackName }).promise();
+
+    // Check if the toolkit stack exists
+    if (!response.Stacks || response.Stacks.length === 0) {
+      throw new Error(`Toolkit stack ${toolkitStackName} not found`);
+    }
+
+    // Check if we can retrieve the bucketName from the output
+    const bucketNameOutput = response.Stacks[0].Outputs?.find(
+      output => output.OutputKey === 'BucketName',
+    );
+    if (!bucketNameOutput || !bucketNameOutput.OutputValue) {
+      throw new Error(`Unable to locate the Staging BucketName output value in the toolkit stack ${toolkitStackName}`);
+    }
+
+    // Check if the bucket exists and is accessible
+    try {
+      await s3.headBucket({ Bucket: bucketNameOutput.OutputValue }).promise();
+      return true;
+    } catch (s3Error) {
+      debug(`Bucket ${bucketNameOutput.OutputValue} either does not exist, or is not accessible: ${s3Error}`);
+      return false;
+    }
+  } catch (e) {
+    debug(`Error checking for staging bucket in toolkit stack: ${e}`);
+    return false;
+  }
+}
+
+async function _bootstrapVersion(sdk: ISDK, stackName: string): Promise<number> {
+
+  try {
+    // get the toolkit stack
+    const cfn = sdk.cloudFormation();
+    const response = await cfn.describeStacks({ StackName: stackName }).promise();
```

Comment:
This now takes 2 `describeStacks`, one here and one for the `_stagingBucketInToolkitStack` call. That seems unnecessary duplication of API calls.

By the way, didn't Eli put these as methods on the `BootstrapStack` class?

Reply:
> 
> By the way, didn't Eli put these as methods on the `BootstrapStack` class?

They were methods on the `Deployments` stack, however he modified publishAssets function to take this `allowCrossAccount` flag, which is also used in `deploy-stack` and `template-body-parameters`. I figured it's better to do this check everywhere rather than just deployments and only pass a true value here